Integrate dropGem with backend

okg00dbye opened the pull request:
Changes:
- Add dropGem method in GemProvider
- Change protobuf definitions of DropResponse and DropRequest
-- DropRequest takes in GemMessage
-- Added `bytes Photo` to Gem and GemMessage
- Update backend logic to fit protobuf
- Use injected dropGem method in DropGem.vue
- Link the different pages to set the different properties of the draft gem


TODO:
- test this (haven't tested the response, imma need to go and debug my setup for a while)
- replace hard-coded generated gem id
- replace hard-coded user's location coordinates
- Make camera / photos buttons work so that user can set the photo (btw, how do we test this on comp?)

Thanks @sebastiantoh for the guidance!

okg00dbye committed:
Add dropGem method and injection key in GemProvider

okg00dbye committed:
Inject method into DropGem.vue

okg00dbye committed:
Update DropRequest and DropResponse

okg00dbye committed:
Remove boolean from DropResponse

booleans arent a type in protobuf :0

okg00dbye committed:
Update backend dropGem methods

okg00dbye committed:
Update example in main.go

okg00dbye committed:
Run golangci lint fix

okg00dbye committed:
Update GemProvider

okg00dbye committed:
Fix typo in GemMap

okg00dbye committed:
Update method for DropGem.vue

okg00dbye committed:
Move dropGem button from AttachMessage to DropGem

okg00dbye committed:
Link setting receiver id from SelectFriend to DropGem

okg00dbye committed:
[cleanup] simplify setReceiver method

sebastiantoh commented:
so sorry, but for the providers, it's likely that you'll need to change them! we're using vuex store instead here (https://github.com/hansel-app/hansel/pull/43).

I can help make the changes right before this PR is merged!

okg00dbye committed:
Add photo (bytes) property to gem protobuf

okg00dbye committed:
Link set message from AttachMessage to DropGem

okg00dbye committed:
Merge branch 'master' into jl/integrate-drop-gem

okg00dbye committed:
Typo in GemProvider

okg00dbye committed:
Update GemColorCarousel to iterate through GemColor

okg00dbye committed:
Link setting gem color to DropGem

okg00dbye committed:
Remove extra DropGem button

okg00dbye committed:
Receive photo from GemMessage in backend

ming-00 commented on api-specs/gem.proto:
```diff
@@ -31,14 +31,24 @@ message Gem {
+}
+
+message GemMessage {
+  string message = 2;
```
Starting from 2 and missing 8!

okg00dbye replied:
ty!

okg00dbye committed:
Fix non-consecutive proto indexes

sebastiantoh commented on api-specs/gem.proto:
```diff
@@ -31,14 +31,24 @@ message Gem {
   string message = 1;
+  double latitude = 2;
+  double longitude = 3;
+  int64 creator_id = 4;
```
hmm, I think we probably don't need creator_id if we can retrieve the person who sent the request from the backend's context?

okg00dbye replied:
removed

sebastiantoh commented on backend/internal/adapters/primary/handlers/internal/gemservice.go:
```diff
@@ -25,14 +26,27 @@ func NewGemService(repository gems.Repository) *gemService {
+	gemMessage := r.GetGemMessage()
+	gem := gems.Gem{
+		// TODO: generate ID here, not sure how.
+		ID:         123,
```
Not 100% certain, but is it necessary to provide all the fields? I think we can probably omit the ID. Likewise for created_at since these will be populated by the db

okg00dbye replied:
removed

sebastiantoh commented on backend/internal/adapters/secondary/repositories/gem.go:
```diff
@@ -51,18 +51,16 @@ func (r *gemRepository) GetPendingCollectionByUser(userId int64) ([]gems.Gem, er
-		"message": message,
-	}).Returning(goqu.T("gems").All()).ToSQL()
+func (r *gemRepository) Add(gem *gems.Gem) (string, error) {
+	sql, _, _ := qb.Insert("gems").ToSQL()
```
hmm, does this work? I don't think this statement actually inserts the gem haha

okg00dbye replied:
HAHA oops, updated

sebastiantoh commented on backend/internal/adapters/secondary/repositories/gem.go:
```diff
@@ -51,18 +51,16 @@ func (r *gemRepository) GetPendingCollectionByUser(userId int64) ([]gems.Gem, er
-	return &gem, nil
+	// Thinking if we should return id of the inserted gem here.
+	// Feel free to overwrite
+	return "Success", nil
```
yep, I think returning id is good!

okg00dbye replied:
done

okg00dbye committed:
Merge branch 'master' into jl/integrate-drop-gem

okg00dbye committed:
Resolve comments

remove fields in gems.Gem
actually insert gem
remove creator_id

okg00dbye committed:
Update dropGem to return int64 dropped_gem_id

okg00dbye commented:
> so sorry, but for the providers, it's likely that you'll need to change them! we're using vuex store instead here (#43).
> 
> I can help make the changes right before this PR is merged!

Yeah I think I lost the dropGem after updating from master hehe, do you want to help add it back?

sebastiantoh commented:
> > so sorry, but for the providers, it's likely that you'll need to change them! we're using vuex store instead here (#43).
> > I can help make the changes right before this PR is merged!
> 
> Yeah I think I lost the dropGem after updating from master hehe, do you want to help add it back?

Ok, sure! 

P.s. I'll likely only do it tomorrow afternoon though

okg00dbye committed:
Add dropGem to gem store

okg00dbye committed:
Update DropGem to use store

okg00dbye commented:
> > Yeah I think I lost the dropGem after updating from master hehe, do you want to help add it back?
> 
> Ok, sure!
> 
> P.s. I'll likely only do it tomorrow afternoon though

churned something out, not sure what i'm writing anymore, can take a look~

sebastiantoh commented:
> > > Yeah I think I lost the dropGem after updating from master hehe, do you want to help add it back?
> > 
> > 
> > Ok, sure!
> > P.s. I'll likely only do it tomorrow afternoon though
> 
> churned something out, not sure what i'm writing anymore, can take a look~

thanks, I think it looks good. In hindsight, I felt like maybe we should have kept the providers since dropGem doesn't necessarily need to mutate the state.

let me test this pr!

sebastiantoh committed:
Shift camelToSnake mapper to query builder

sebastiantoh commented:
I think it works great except for the following:

- photos column is not defined in Postgesql
  - Not sure how we want to do this, do we want to store the files in Postgres, or on the server itself?
  - For simplicity, I propose adding the following columns to the gems table: `(attachment_name VARCHAR, attachment BYTEA)`

- zero values are being inserted into DB as well (e.g. creator_id = 0, created_at = 0001-01-01T00:00:00Z
  - Can think of 2 alternatives for this: individually specify the fields using records (https://doug-martin.github.io/goqu/docs/inserting.html#insert-record) or add struct tags `goqu:"defaultifempty"`

- Probably need to send over the current lat/long as well.
  - For this, I think we probably need another store that stores user info (see my comment in your other PR)

- need to extract out userid from context and add it to gem struct

thoughts?

P.s. I shifted the cameltosnake mapper to the querybuilder instead. But I forgot to test if it affected the other queries (I think it will, will test later)

okg00dbye commented:
> I think it works great except for the following:
> 
> * photos column is not defined in Postgesql
>   
>   * Not sure how we want to do this, do we want to store the files in Postgres, or on the server itself?
>   * For simplicity, I propose adding the following columns to the gems table: `(attachment_name VARCHAR, attachment BYTEA)`

sgtm


> * zero values are being inserted into DB as well (e.g. creator_id = 0, created_at = 0001-01-01T00:00:00Z
>   
>   * Can think of 2 alternatives for this: individually specify the fields using records (https://doug-martin.github.io/goqu/docs/inserting.html#insert-record) or add struct tags `goqu:"defaultifempty"`

is creator_id the userid we are extracting from context?

> * Probably need to send over the current lat/long as well.
>   
>   * For this, I think we probably need another store that stores user info (see my comment in your other PR)

yup! will see if i can update this part in the other PR

> * need to extract out userid from context and add it to gem struct
> 
> thoughts?
> 
> P.s. I shifted the cameltosnake mapper to the querybuilder instead. But I forgot to test if it affected the other queries (I think it will, will test later)

how do you wake up so early hahaha

sebastiantoh committed:
Use struct tags instead of mapping

sebastiantoh committed:
Add attachment column to db

sebastiantoh committed:
Rename photo to attachment

sebastiantoh committed:
Retrieve and set UserID as CreatorId

sebastiantoh committed:
Add goqu  struct tags

sebastiantoh committed:
Fix LastInsertId error

sebastiantoh commented:
okie, I think this PR can be reviewed! The insertions should work. I'll work on fully integrating the frontend together with https://github.com/hansel-app/hansel/pull/30

ianyong commented on api-specs/gem.proto:
```diff
@@ -31,15 +31,24 @@ message Gem {
 message DropResponse {
-  int64 id = 1;
-  string message = 2;
+  int64 dropped_gem_id = 2;
```
Is this meant to start from 2?

okg00dbye replied:
fixed tq

ianyong commented on backend/internal/core/domain/gems/gem.go:
```diff
@@ -16,13 +16,14 @@ const (
-	ReceiverId int64
-	ReceivedAt *time.Time
-	Color      GemColor
+	ID         int64      `db:"id" goqu:"skipinsert"`
+	Message    string     `db:"message"`
+	Latitude   float64    `db:"latitude"`
+	Longitude  float64    `db:"longitude"`
+	CreatorId  int64      `db:"creator_id"`
+	CreatedAt  time.Time  `db:"created_at" goqu:"skipinsert"`
+	ReceiverId int64      `db:"receiver_id"`
+	ReceivedAt *time.Time `db:"received_at"`
+	Color      GemColor   `db:"color"`
+	Attachment []byte     `db:"attachment"`
```
Technically results in the domain models being coupled to the database, but oh well.

sebastiantoh replied:
hmm, i would like to think of it as the database using the domain model so the dependency is still inwards (logically)

ianyong commented on frontend/src/pages/DropGem/DropGem.vue:
```diff
@@ -47,6 +60,28 @@ export default defineComponent({
+      this.draftGem.setColor(value);
+    },
+    dropMyGem() {
+      console.log("called dropMyGem");
+      const dropGem = () => this.store.dispatch("dropGem", this.draftGem);
+      const response = ref<number>(-1);
+      console.log(this.draftGem);
+      dropGem()
+        .then((resp) => {
+          response.value = resp;
+        })
+        .catch((err) => console.log(err, "Failed to drop gem"));
+      console.log(response.value);
```
Should these `console.log` statements be removed?

okg00dbye replied:
removed

ianyong commented on frontend/src/store/modules/gems.ts:
```diff
@@ -56,12 +60,16 @@ const protoGemToGemMapper = (protoGem: ProtoGem): Gem => {
 const gemsModule: Module<GemsState, RootState> = {
   state: {
     gemsPendingCollection: [],
+    droppedGemId: -1,
```
What is this for? Also, can we avoid magic values?

okg00dbye replied:
changed -1 to undefined
it's not for much, seems like can't define action without mutating state
https://stackoverflow.com/questions/52630866/vuex-actions-that-do-not-need-to-commit-a-mutation

okg00dbye committed:
Resolve comments

Fix wrong starting index in gem.proto
Remove console.log statements
Change magic number to undefined

ianyong reviewed (approved):
LGTM! Thanks!
